### ISSUE TITLE
Mismatch between the type derived from the ```lean``` query and the declared interface on virtual fields

### ISSUE THREAD

**Issue by nikzanda** [Automattic/mongoose#14649]
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure the bug has not already been reported

### Mongoose version

8.4.1

### Node.js version

20.10.0

### MongoDB server version

6.0.2

### Typescript version (if applicable)

5.4.5

### Description

reproduction link: https://stackblitz.com/edit/stackblitz-starters-4vhsz6?file=src%2Findex.ts

As you can see from the reproduction link, the user model interface contains both the fields saved to the database and getters and virtuals. In the options, I added `toJSON: { virtuals: true }` and `toObject: { virtuals: true }` to populate the virtual fields when I query on user. When I do lean queries, the returned type matches that of the interface but this is wrong, because the getters and virtuals are not populated and for typescript they are not undefined fields.

![image](https://github.com/Automattic/mongoose/assets/71180446/8ba95a7f-5b31-4833-b127-9bb05b8f17f4)
![Screenshot 2024-06-04 at 08 50 13](https://github.com/Automattic/mongoose/assets/71180446/ab9cfaaa-6527-41d3-a89e-e0492986b3ae)
![Screenshot 2024-06-04 at 08 52 37](https://github.com/Automattic/mongoose/assets/71180446/cad978ad-b8c0-4c26-839c-f7c14c5ca937)

I’m not interested in populating the virtual fields if the query is lean, however, the TypeScript definition should return undefined for unpopulated fields.

### Steps to Reproduce

Clone the repository locally, compile the TypeScript code and run the command:
```sh
npx tsc
node --experimental-specifier-resolution=node dist/index.js
```

### Expected Behavior

The virtual fields should not exist in the "leanUser" variable or be of type `type | undefined`

**Comment by vkarpov15** [comment]
The issue is that you're defining an `IUser` interface that lists `myVirtualField` as a property. If you define the raw doc interface (`IUser` in your case), it should define the document as it is stored in MongoDB, which excludes any virtuals. Define a separate `IUserVirtuals` interface as follows.

```ts
import {
  Schema,
  Connection,
  Model,
  SchemaOptions,
  SchemaTypes,
  model
} from 'mongoose';

export interface IUser {
  id: string; // getter
  name: string;
  createdAt: Date;
  updatedAt: Date;
}

export interface IUserVirtuals {
  myVirtualField: string; // virtual
}

type UserModelType = Model<IUser, {}, {}, IUserVirtuals>;

export type UserInstance = InstanceType<UserModelType>;

const schema = new Schema<IUser, UserModelType, {}, {}, IUserVirtuals>({ name: String }, { timestamps: true })

const UserModel = model<IUser, UserModelType>('User', schema);

UserModel.findOne().lean().orFail().then(doc => {
  doc.myVirtualField;
});

```

**Comment by vkarpov15** [comment]
Yes, you should define the id field in the IUserVirtuals interface in Mongoose 8
